server/market/orderrouter.go

```diff
@@ -0,0 +1,577 @@
+		assets:   cfg.Assets,
+		tunnels:  cfg.Markets,
+		mbBuffer: cfg.MarketBuyBuffer,
+	}
+	cfg.AuthManager.Route(msgjson.LimitRoute, router.handleLimit)
+	cfg.AuthManager.Route(msgjson.MarketRoute, router.handleMarket)
+	cfg.AuthManager.Route(msgjson.CancelRoute, router.handleCancel)
+	return router
+}
+
+// handleLimit is the handler for the 'limit' route. This route accepts a
+// msgjson.Limit payload, validates the information, constructs an
+// order.LimitOrder and submits it to the epoch queue.
+func (r *OrderRouter) handleLimit(user account.AccountID, msg *msgjson.Message) *msgjson.Error {
+	limit := new(msgjson.Limit)
+	err := json.Unmarshal(msg.Payload, limit)
+	if err != nil {
+		return msgjson.NewError(msgjson.RPCParseError, "error decoding 'limit' payload")
+	}
+
+	rpcErr := r.verifyAccount(user, limit.AccountID, limit)
+	if rpcErr != nil {
+		return rpcErr
+	}
+
+	tunnel, coins, sell, rpcErr := r.extractMarketDetails(&limit.Prefix, &limit.Trade)
+	if rpcErr != nil {
+		return rpcErr
+	}
+
+	// Check that OrderType is set correctly
+	if limit.OrderType != msgjson.LimitOrderNum {
+		return msgjson.NewError(msgjson.OrderParameterError, "wrong order type set for limit order")
+	}
+
+	valSum, spendSize, utxos, rpcErr := r.checkPrefixTrade(user, tunnel, coins, &limit.Prefix, &limit.Trade, true)
+	if rpcErr != nil {
+		return rpcErr
+	}
+
+	// Check that the rate is non-zero and obeys the rate step interval.
+	if limit.Rate == 0 {
+		return msgjson.NewError(msgjson.OrderParameterError, "rate = 0 not allowed")
+	}
+	if limit.Rate%coins.quote.RateStep != 0 {
+		return msgjson.NewError(msgjson.OrderParameterError, "rate not a multiple of ratestep")
+	}
+
+	// Calculate the fees and check that the utxo sum is enough.
+	swapVal := limit.Quantity
+	if !sell {
+		swapVal = matcher.BaseToQuote(limit.Rate, limit.Quantity)
+	}
+	reqVal := requiredFunds(swapVal, spendSize, coins.funding)
+	if valSum < reqVal {
+		return msgjson.NewError(msgjson.FundingError,
+			fmt.Sprintf("not enough funds. need at least %d, got %d", reqVal, valSum))
+	}
+
+	// Check time-in-force
+	if !(limit.TiF == msgjson.StandingOrderNum || limit.TiF == msgjson.ImmediateOrderNum) {
+		return msgjson.NewError(msgjson.OrderParameterError, "unknown time-in-force")
+	}
+
+	// Create the limit order
+	serverTime := time.Now().UTC()
+	lo := &order.LimitOrder{
+		MarketOrder: order.MarketOrder{
+			Prefix: order.Prefix{
+				AccountID:  user,
+				BaseAsset:  limit.Base,
+				QuoteAsset: limit.Quote,
+				OrderType:  order.LimitOrderType,
+				ClientTime: time.Unix(int64(limit.ClientTime), 0).UTC(),
+				ServerTime: serverTime,
+			},
+			UTXOs:    utxos,
+			Sell:     sell,
+			Quantity: limit.Quantity,
+			Address:  limit.Address,
+		},
+		Rate:  limit.Rate,
+		Force: order.StandingTiF,
+	}
+
+	// Send the order to the epoch queue.
+	oRecord := &orderRecord{
+		order: lo,
+		req:   limit,
+		msgID: msg.ID,
+	}
+	tunnel.SubmitOrderAsync(oRecord)
+	r.respondOrder(oRecord)
```

Comment:
I've separated the signing/response from the epoch submission, but if we move this `respondOrder` method to `Market` and do it in the main loop, that will solve the epoch synchronization issue.

Reply:
Presently, Market has access to the auth manager via the swapper, so it can implement `respondOrder`.  Let's try that.